tests/planned_maintenance_test.py

```diff
@@ -0,0 +1,183 @@
+from datetime import datetime, timedelta
+from unittest.mock import Mock
+
+import pytest
+
+from zino.planned_maintenance import PlannedMaintenances
+from zino.state import ZinoState
+from zino.statemodels import (
+    AlarmEvent,
+    DeviceMaintenance,
+    EventState,
+    Port,
+    PortStateEvent,
+    PortStateMaintenance,
+    ReachabilityEvent,
+)
+
+
+def test_should_start_with_no_planned_maintenances(pms):
+    assert len(pms) == 0
+
+
+def test_pm_should_be_gettable_by_id(pms, active_pm):
+    assert pms[active_pm.id] == active_pm
+
+
+class TestGetPlannedMaintenances:
+    def test_get_started_planned_maintenances(self, pms, active_pm, ended_pm, old_pm):
+        pms.last_run = datetime.now() - timedelta(hours=1)
+        recent_pm = pms.create_planned_maintenance(
+            start_time=datetime.now() - timedelta(minutes=1),
+            end_time=datetime.now() + timedelta(days=1),
+            pm_class=DeviceMaintenance,
+            match_type="str",
+            match_expression="hello",
+            match_device="device",
+        )
+        started_pms = pms.get_started_planned_maintenances(now=datetime.now())
+        assert recent_pm in started_pms
+        assert active_pm not in started_pms
+        assert ended_pm not in started_pms
+        assert old_pm not in started_pms
+
+    def test_get_ended_planned_maintenances(self, pms, active_pm, ended_pm, old_pm):
+        pms.last_run = datetime.now() - timedelta(hours=1)
+        ended_pms = pms.get_ended_planned_maintenances(now=datetime.now())
+        assert ended_pm in ended_pms
+        assert old_pm not in ended_pms
+        assert active_pm not in ended_pms
+
+    def test_get_active_planned_maintenances(self, pms, active_pm, ended_pm, old_pm):
+        active_pms = pms.get_active_planned_maintenances(datetime.now())
+        assert active_pm in active_pms
+        assert ended_pm not in active_pms
+        assert old_pm not in active_pms
+
+    def test_get_old_planned_maintenances(self, pms, active_pm, ended_pm, old_pm):
+        old_pms = pms.get_old_planned_maintenances(now=datetime.now())
+        assert old_pm in old_pms
+        assert active_pm not in old_pms
+        assert ended_pm not in old_pms
```

Comment:
Same thing here: It's best to keep the number of asserts per test limited - at least not assert multiple different behaviors in a single test.

Reply:
fixed